Add systemd sd_notify()

nrgaway opened the pull request:
A systemd unit file that has the 'Type' set to 'notify' enables the ability for systemd to notify any unit that 'Requires' it when a call is made to sd_nofify().  The unit file that requires to be notified will not start until systemd recieves a 'READY=1' message which allows services that rely on another service to be fully initialized to wait preventing race conditions.
    
The kvmchand service unit file service 'Type' was changed from 'simple' to 'notify' and dropins were added for qubes-db.service and qubes-qrexec.service to `Require` the kvmchand.service unit to force them to wait until kvmchan initialization is complete since the services were starting too early which caused fatal communication errors that could not recover.

nrgaway committed:
Add systemd sd_notify()

nrgaway committed:
Moved systemd sd_notify() from vfio.c to localhandler.c

A systemd unit file that has the 'Type' set to 'notify' enables the ability for
systemd to notify any unit that 'Requires' it when a call is made to
sd_nofify().  The unit file that requires to be notified will not start until
systemd recieves a 'READY=1' message which allows services that rely on another
service to be fully initialized to wait preventing race conditions.

The kvmchand service unit file service 'Type' was changed from 'simple' to
'notify' and dropins were added for qubes-db.service and qubes-qrexec.service
to `Require` the kvmchand.service unit to force them to wait until kvmchan
initialization is complete since the services were starting too early which
caused fatal communication errors that could not recover.

nrgaway commented:
Following are the _kvmchand_ parts of a guest log when booting without sd_notify built into _kvmchand_.
```make
You are in emergency mode. After logging in, type "journalctl -xb" to view
system logs, "systemctl reboot" to reboot, "systemctl default" or "exit"
to boot into default mode.

Cannot open access to console, the root account is locked.
See sulogin(8) man page for more details.

#Aug 23 22:33:28 localhost systemd-tmpfiles[476]: /usr/lib/tmpfiles.d/qubes-core-agent-linux.conf:1: Line references path below legacy directory /var/run/, updating /var/run/tinyproxy-updates → /run/tinyproxy-updates; please update the tmpfiles.d/ drop-in file accordingly.
#Aug 23 22:33:28 localhost systemd-tmpfiles[476]: /etc/tmpfiles.d/qubes-gpg-split.conf:1: Line references path below legacy directory /var/run/, updating /var/run/qubes-gpg-split → /run/qubes-gpg-split; please update the tmpfiles.d/ drop-in file accordingly.
Aug 23 22:33:29 localhost qubesdb-daemon[493]: FATAL: vchan initialization failed
Aug 23 22:33:29 localhost systemd[1]: Starting KVM vchan daemon...
Aug 23 22:33:29 localhost systemd[1]: Started KVM vchan daemon.
Aug 23 22:33:29 localhost systemd[1]: Starting Qubes DB agent...
Aug 23 22:33:29 localhost systemd[1]: qubes-db.service: Main process exited, code=exited, status=1/FAILURE
Aug 23 22:33:29 localhost systemd[1]: qubes-db.service: Failed with result 'exit-code'.
Aug 23 22:33:29 localhost systemd[1]: Failed to start Qubes DB agent.
#Aug 23 22:33:29 localhost systemd[1]: Starting Init Qubes Services settings...
Aug 23 22:33:29 localhost kvmchand[506]: [INFO] daemon/localhandler.c:580: guest_main: sd_notify

## 1st RAM MEMORY DEVICE
## 00:10.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)
## 00:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
Aug 23 22:33:29 localhost kvmchand[505]: [INFO] daemon/vfio.c:1205: Got ivshmem device: 0000:00:10.0
########

## 2nd RAM MEMORY DEVICE
## 00:11.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)
## 00:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
Aug 23 22:33:29 localhost kvmchand[505]: [INFO] daemon/vfio.c:1205: Got ivshmem device: 0000:00:11.0
########

Aug 23 22:33:29 localhost kvmchand[505]: [WARN] daemon/vfio.c:1221: Some ivshmem devices aren't bound to vfio-pci. Attempting to bind...
Aug 23 22:33:29 localhost kernel: vfio-pci 0000:00:10.0: Adding to iommu group 0  ## 1st RAM MEMORY DEVICE
Aug 23 22:33:29 localhost kernel: vfio-pci 0000:00:10.0: Adding kernel taint for vfio-noiommu group on device  ## 1st RAM MEMORY DEVICE
Aug 23 22:33:29 localhost kernel: vfio-pci 0000:00:11.0: Adding to iommu group 1  ## 2nd RAM MEMORY DEVICE
Aug 23 22:33:29 localhost kernel: vfio-pci 0000:00:11.0: Adding kernel taint for vfio-noiommu group on device  ## 2nd RAM MEMORY DEVICE
Aug 23 22:33:29 localhost kvmchand[505]: [WARN] daemon/vfio.c:1236: Successfully bound ivshmem devices.
Aug 23 22:33:29 localhost kernel: vfio-pci 0000:00:10.0: vfio-noiommu device opened by user (kvmchand:505)  ## 1st RAM MEMORY DEVICE
Aug 23 22:33:29 localhost kernel: vfio-pci 0000:00:11.0: vfio-noiommu device opened by user (kvmchand:505)  ## 2nd RAM MEMORY DEVICE
Aug 23 22:33:29 localhost kvmchand[505]: [INFO] daemon/vfio.c:993: Successfully connected to host daemon.
#Aug 23 22:33:29 localhost qubes-sysinit.sh[565]: Failed connect to local daemon
#Aug 23 22:33:29 localhost qubes-sysinit.sh[567]: Failed connect to local daemon
#Aug 23 22:33:29 localhost qubes-sysinit.sh[569]: Failed connect to local daemon
#Aug 23 22:33:29 localhost qubes-sysinit.sh[571]: Failed connect to local daemon
#Aug 23 22:33:29 localhost qubes-random-seed.sh[579]: Failed connect to local daemon
#Aug 23 22:33:29 localhost systemd[1]: Finished Init Qubes Services settings.
#Aug 23 22:33:29 localhost systemd[1]: Starting Initialize and mount /rw and /home...
#Aug 23 22:33:29 localhost systemd[1]: Starting Adjust root filesystem size...
#Aug 23 22:33:29 localhost systemd[1]: Finished Initialize and mount /rw and /home.
#Aug 23 22:33:29 localhost systemd[1]: Finished Adjust root filesystem size.
Aug 23 22:33:29 localhost systemd[1]: Started Emergency Shell.
Aug 23 22:33:29 localhost systemd[1]: Reached target Emergency Mode.
#Aug 23 22:33:29 localhost systemd[1]: Starting Early Qubes VM settings...
#Aug 23 22:33:29 localhost qubes-early-vm-config.sh[915]: Failed connect to local daemon
#Aug 23 22:33:29 localhost systemd[1]: Finished Early Qubes VM settings.
#Aug 23 22:33:29 localhost systemd-tmpfiles[911]: /usr/lib/tmpfiles.d/qubes-core-agent-linux.conf:1: Line references path below legacy directory /var/run/, updating /var/run/tinyproxy-updates → /run/tinyproxy-updates; please update the tmpfiles.d/ drop-in file accordingly.
#Aug 23 22:33:29 localhost systemd-tmpfiles[911]: /etc/tmpfiles.d/qubes-gpg-split.conf:1: Line references path below legacy directory /var/run/, updating /var/run/qubes-gpg-split → /run/qubes-gpg-split; please update the tmpfiles.d/ drop-in file accordingly.
#Aug 23 22:33:29 localhost systemd-udevd[529]: vda1: Process '/usr/lib/qubes/udev-block-add-change' failed with exit code 1.
#Aug 23 22:33:29 localhost systemd[1]: Mounted /rpm.
#Aug 23 22:33:29 localhost systemd[1]: Mounted /share.
#Aug 23 22:33:29 localhost systemd-udevd[536]: vda2: Process '/usr/lib/qubes/udev-block-add-change' failed with exit code 1.
Aug 23 22:33:47 localhost kvmchand[506]: [INFO] daemon/localhandler.c:620: Client connected! fd: 8
Aug 23 22:33:47 localhost kvmchand[505]: [INFO] daemon/vfio.c:1117: resp from host: err: 0, ret: 2
#Aug 23 22:33:47 localhost systemd[1]: Starting Qubes misc post-boot actions...
Aug 23 22:33:47 localhost systemd[1]: Starting Qubes remote exec agent...
Aug 23 22:33:47 localhost qrexec-agent[1028]: 2020-08-23 22:33:47.413 qrexec-agent[1028]: qrexec-agent.c:376:init: qrexec-agent.init: ctrl_vchan call
Aug 23 22:33:47 localhost kvmchand[506]: [INFO] daemon/localhandler.c:620: Client connected! fd: 9
Aug 23 22:33:47 localhost kvmchand[505]: [INFO] daemon/vfio.c:1117: resp from host: err: 0, ret: 3
Aug 23 22:33:47 localhost kernel: vfio-pci 0000:00:10.0: EDR: ACPI event 0x1 received  ## 1st RAM MEMORY DEVICE
Aug 23 22:33:47 localhost kernel: vfio-pci 0000:00:11.0: EDR: ACPI event 0x1 received  ## 2nd RAM MEMORY DEVICE

## 3rd RAM MEMORY DEVICE
## 00:12.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)
## 00:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
Aug 23 22:33:47 localhost kernel: pci 0000:00:12.0: [1af4:1110] type 00 class 0x050000
Aug 23 22:33:47 localhost kernel: pci 0000:00:12.0: reg 0x10: [mem 0x00000000-0x000000ff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:12.0: reg 0x14: [mem 0x00000000-0x00000fff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:12.0: reg 0x18: [mem 0x00000000-0x00003fff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:12.0: BAR 2: assigned [mem 0x800134000-0x800137fff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:12.0: BAR 1: assigned [mem 0xc844e000-0xc844efff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:12.0: BAR 0: assigned [mem 0xc844f000-0xc844f0ff]
########

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]

Aug 23 22:33:47 localhost kernel: vfio-pci 0000:00:12.0: Adding to iommu group 2  ## 3rd RAM MEMORY DEVICE
Aug 23 22:33:47 localhost kernel: vfio-pci 0000:00:12.0: Adding kernel taint for vfio-noiommu group on device  ## 3rd RAM MEMORY DEVICE

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]

#Aug 23 22:33:47 localhost systemd[1]: Started Serial Getty on ttyS0.

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: bridge window [mem 0x00100000-0x000fffff 64bit pref] to [bus 01] add_size 200000 add_align 100000
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: BAR 15: assigned [mem 0x800200000-0x8003fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]

Aug 23 22:33:47 localhost kernel: pci_bus 0000:02: Allocating resources
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: bridge window [mem 0x00100000-0x000fffff 64bit pref] to [bus 02] add_size 200000 add_align 100000
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: BAR 15: assigned [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:47 localhost kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: vfio-pci 0000:00:12.0: vfio-noiommu device opened by user (kvmchand:514)  ## 3rd RAM MEMORY DEVICE
Aug 23 22:33:51 localhost.localdomain kvmchand[506]: [INFO] daemon/localhandler.c:328: fds: {10, 11, 12, 13, 14}
Aug 23 22:33:51 localhost.localdomain systemd[1]: Started Qubes DB agent.
#Aug 23 22:33:51 localhost.localdomain nm-dispatcher[1297]: /usr/lib/qubes/qubes-setup-dnat-to-ns: line 17: /var/run/qubes/qubes-ns: No such file or directory
Aug 23 22:33:51 localhost.localdomain kvmchand[506]: [INFO] daemon/localhandler.c:328: fds: {10, 11, 12, 13, 14}
Aug 23 22:33:51 localhost.localdomain qrexec-agent[1028]: 2020-08-23 22:33:51.424 qrexec-agent[1028]: qrexec-agent.c:378:init: qrexec-agent.init: ctrl_vchan returned
Aug 23 22:33:51 localhost.localdomain systemd[1]: Started Qubes remote exec agent.
#Aug 23 22:33:51 localhost.localdomain systemd[1]: Reached target Multi-User System.
#Aug 23 22:33:51 localhost.localdomain kernel: fbcon: Taking over console
Aug 23 22:33:51 localhost.localdomain kvmchand[506]: [INFO] daemon/localhandler.c:620: Client connected! fd: 10

## 4th RAM MEMORY DEVICE
## 00:13.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)
## 00:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:13.0: [1af4:1110] type 00 class 0x050000
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:13.0: reg 0x10: [mem 0x00000000-0x000000ff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:13.0: reg 0x14: [mem 0x00000000-0x00000fff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:13.0: reg 0x18: [mem 0x00000000-0x0003ffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:13.0: BAR 2: assigned [mem 0x800140000-0x80017ffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:13.0: BAR 1: assigned [mem 0xc8450000-0xc8450fff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:13.0: BAR 0: assigned [mem 0xc8451000-0xc84510ff]
########

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kvmchand[505]: [INFO] daemon/vfio.c:1117: resp from host: err: 0, ret: 4
Aug 23 22:33:51 localhost.localdomain kernel: vfio-pci 0000:00:13.0: Adding to iommu group 3  ## 4th RAM MEMORY DEVICE
Aug 23 22:33:51 localhost.localdomain kernel: vfio-pci 0000:00:13.0: Adding kernel taint for vfio-noiommu group on device  ## 4th RAM MEMORY DEVICE

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci_bus 0000:01: Allocating resources
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci_bus 0000:02: Allocating resources
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0: PCI bridge to [bus 01]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [io  0xd000-0xdfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0xc8200000-0xc83fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:02.0:   bridge window [mem 0x800200000-0x8003fffff 64bit pref]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0: PCI bridge to [bus 02]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [io  0xc000-0xcfff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0xc8000000-0xc81fffff]
Aug 23 22:33:51 localhost.localdomain kernel: pci 0000:00:03.0:   bridge window [mem 0x800400000-0x8005fffff 64bit pref]

Aug 23 22:33:55 localhost.localdomain kernel: vfio-pci 0000:00:13.0: vfio-noiommu device opened by user (kvmchand:514)  ## 4th RAM MEMORY DEVICE
Aug 23 22:33:55 localhost.localdomain kvmchand[506]: [INFO] daemon/localhandler.c:328: fds: {11, 12, 13, 14, 15}
```

shawnanastasio commented:
> I apologize for the late response; I have been sick for the last week :(

No worries at all, I hope you are feeling better!

> I tested with _sd_notify_ disabled and while _qubes-db_ and _qrexec_ do end up connecting, the system starts in emergency mode and I have had failures when attempting to run `qvm-run` from the host. Until further investigation I am guessing it's due to other services that depend on either _qubes-db_ or _qrexec_ being fully initialized as well.
> 
> At the moment, I prefer to use _sd_notify_ to ensure _kvmchand_ is fully initialized before any other _Qubes_ modules attempt to load which should prevent any race conditions. If you prefer not to have the systemd notify built-in by default or within _libkvmchan_ we can either change the Makefile SYSTEMD option from `SYSTEMD ?= 1` to `SYSTEMD ?= 0` or I can just apply a patch at build time.

From what I can tell, this is happening because qubesdb is spawned immediately after kvmchand and attempts to connect to the kvmchand socket before localhandler has set it up. The current library behavior is to simply return a failure when the socket connection fails, and that failure is caught by qubesdb.

I'm a bit hesitant to include systemd-specific workarounds for what could be considered an issue with libkvmchan, but in this case it does seem like sd_notify is one of the cleaner solutions. If you could change the default value of `SYSTEMD` to 0 and set it manually it in your build scripts, that would be ideal.

> To provide a better understanding of how _kvmchand_ fits into the boot sequence I have provided an outline below.
> 
> The _kvmchand service_ is wanted-by the systemd _sysinit.target_, which means it runs early. It also contains an `ExecStartPre` statement which makes sure VIFO modulus are loaded, `enable_unsafe_noiommu_mode` is enabled and sets a flag to
> indicate the system is running within a _KVM_ guest. FYI, The _qubes-db service_ type is also 'notify' and is also wanted by _sysinit.target_. Note that I plan to move the code that sets the hypervisor type to a systemd unit file named `hypervisor.service` which will call a script to set the hypervisor type (IE: `touch /var/run/qubes-service/hypervisor-{kvm|xen}`).
> 
> My goal is to be able to build a **single** version of _Qubes_ that use the same packages for both _Xen_ and _KVM_. For instance, the same _Qubes_ virtual machine template image could be used within the current _Qubes_ _Xen_ hypervisor environment or copied to a _KVM_ host and used unmodified.
> 
> To achieve this goal I have created systemd drop-ins for _KVM_ that include a condition `ConditionPathExists=/var/run/qubes-service/hypervisor-kvm`. As it turns out this logic appears flawed though. I thought adding a drop-in based on a condition would prevent the drop-in from being applied if the `hypervior-kvm` file did not exist and the main unit file would still run for Xen based systems, but it seems that the drop-ins just get merged on top of the main systemd unit file and since the condition is false on a _Xen_ based system, the unit will not run.
> 
> Since it does not seem I can add systemd drop-ins to be applied conditionally, I have focused my efforts on using a systemd generator to dynamicly add the _KVM_ drop-ins when a _KVM_ hypervisor is detected. The code is written and the drop-ins
> get created within the `/var/run/systemd/generator.early` directory but do not get applied during boot, but do get applied when re-starting a service, so I need to do further research.
> 
> For completeness I have also included samples of the current configuration files and scripts below.
> [snip]

Unfortunately I'm no expert when it comes to init systems or systemd in particular, but the idea of being able to use the same VM images on both Xen and KVM hosts does sound appealing. If there's anything kvmchand can do to make this easier, I'm open to suggestions (perhaps kvmchand can be made to detect Xen and exit gracefully? I'm not sure if this would simplify any of the service dependency logic.)

nrgaway committed:
Makefile:  Change SYSTEMD default from 1 to 0

Now requires manually set SYSTEMD value to 1 when building libkvmchan to
preserve the build behaviour for non Qubes builds.

nrgaway commented:
> I'm a bit hesitant to include systemd-specific workarounds for what could be considered an issue with libkvmchan, but in this case it does seem like sd_notify is one of the cleaner solutions. If you could change the default value of `SYSTEMD` to 0 and set it manually it in your build scripts, that would be ideal.

Completed.

> Unfortunately I'm no expert when it comes to init systems or systemd in particular, but the idea of being able to use the same VM images on both Xen and KVM hosts does sound appealing. If there's anything kvmchand can do to make this easier, I'm open to suggestions (perhaps kvmchand can be made to detect Xen and exit gracefully? I'm not sure if this would simplify any of the service dependency logic.)

I'm sure I will get the _systemd_ generators working.  I must just be missing something simple as I have created generators in the past.  _Xen_ detection can be handled by the _systemd_ service unit by adding a condition `!ConditionVirtualization="xen"` so _kvmchand_ would not need to implement anything.

shawnanastasio commented:
Thanks!
